Add pprof collection to 100 node scale test

learnitall opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [X] Thanks for contributing!

This PR utilizes a recent change to ClusterLoader2 that allows for [injecting additional measurements](https://github.com/kubernetes/perf-tests/pull/2483) to collect pprofs throughout the duration of the 100 node scale test. These pprofs are saved as artifacts and can be used for debugging.

Pprofs are pulled using the PodPeriodicCommand measurement, which periodically executes `cilium-bugtool --get-pprof` inside a randomly selected cilium-agent. A tar archive containing the profiles is saved to stdout, which is saved by the PodPeriodicCommand measurement and can later be extracted.

```release-note
Scrape pprofs in 100 node scale test workflow for extra debugging information
```

learnitall committed:
ci: Add CL2 module for pulling cilium-agent pprofs

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Collect cilium-agent pprofs during 100 node scale test

This commit adjusts the 100 node scale test to include the
cilium-agent pprofs CL2 module. This will trigger the collection
of cilium-agent pprofs throughout the duration of the test,
assisting in debugging of regressions.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Enable cilium-agent pprof endpoint in 100 node scale test

This commit enables the pprof endpoint in cilium-agent in the 100 node
scale test. This will be used in subsequent commits to collect pprofs
during the test.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

giorio94 commented on .github/workflows/scale-test-node-throughput-gce.yaml:
```diff
@@ -189,6 +190,7 @@ jobs:
             --provider=gce \
             --enable-prometheus-server \
             --testoverrides=./testing/prometheus/not-scrape-kube-proxy.yaml \
+            --testoverrides=./modules.yaml \
```
Should this change (and the one above) reverted based on https://github.com/cilium/cilium/pull/32056#issuecomment-2064227760?

learnitall replied:
Good catch, thank you!

giorio94 commented on contrib/cl2-modules/cilium-agent-pprofs.yaml:
Nit. Other testing related artifacts are stored in `.github/actions/...`. You may want to consider moving this file there as well.

learnitall replied:
I was thinking about this and eventually decided on putting it in `contrib` because I thought everything in `.github/actions` was an actual action, ie contained an `action.yml`, but `.github/actions/aws` and `.github/actions/azure` are great counter-examples. I'll move them :+1:

giorio94 commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -32,6 +32,8 @@ concurrency:
   # renovate: datasource=golang-version depName=go
   go_version: 1.22.2
+  # renovate: datasource=git-refs depName=https://github.com/kubernetes/perf-tests
+  cl2_version: a7c33443c28bbecb53584e3ca26422b3938e5ff2
```
This change is still not enough, because it is missing the definition of the target branch, and it matching the incorrect regex. You can fix it with the following patch:

```diff
diff --git a/.github/renovate.json5 b/.github/renovate.json5
index a7069b29e497..a4a133b97905 100644
--- a/.github/renovate.json5
+++ b/.github/renovate.json5
@@ -452,7 +452,8 @@
         "# renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*?)\\s+.+kernel: '(?<currentValue>.*)'",
         '# renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*)\\s+.+kernel: "(?<currentValue>.*)@(?<currentDigest>sha256:[a-f0-9]+)"',
         '# renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*)\\s+.+image-version: "(?<currentValue>.*)@(?<currentDigest>sha256:[a-f0-9]+)"',
-        '# renovate: datasource=(?<datasource>.*?)\\s+.+kube-image: "(?<depName>.*):(?<currentValue>.*)@(?<currentDigest>sha256:[a-f0-9]+)"'
+        '# renovate: datasource=(?<datasource>.*?)\\s+.+kube-image: "(?<depName>.*):(?<currentValue>.*)@(?<currentDigest>sha256:[a-f0-9]+)"',
+        '# renovate: datasource=(?<datasource>.*?) depName=(?<depName>.*?) branch=(?<currentValue>.*?)\\s+.+_ref: (?<currentDigest>.*)'
       ]
     },
     {
diff --git a/.github/workflows/scale-test-100-gce.yaml b/.github/workflows/scale-test-100-gce.yaml
index ed92e0d5482e..9c664e3b44b3 100644
--- a/.github/workflows/scale-test-100-gce.yaml
+++ b/.github/workflows/scale-test-100-gce.yaml
@@ -32,8 +32,8 @@ concurrency:
 env:
   # renovate: datasource=golang-version depName=go
   go_version: 1.22.2
-  # renovate: datasource=git-refs depName=https://github.com/kubernetes/perf-tests
-  cl2_version: a7c33443c28bbecb53584e3ca26422b3938e5ff2
+  # renovate: datasource=git-refs depName=https://github.com/kubernetes/perf-tests branch=master
+  cl2_ref: a7c33443c28bbecb53584e3ca26422b3938e5ff2
   # Adding k8s.local to the end makes kops happy-
   # has stricter DNS naming requirements.
   test_name: scale-100
@@ -116,7 +116,7 @@ jobs:
         uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         with:
           repository: kubernetes/perf-tests
-          ref: ${{ env.cl2_version }}
+          ref: ${{ env.cl2_ref }}
           persist-credentials: false
           sparse-checkout: clusterloader2
           path: perf-tests
diff --git a/.github/workflows/scale-test-node-throughput-gce.yaml b/.github/workflows/scale-test-node-throughput-gce.yaml
index 1b409f40a592..60b32a58fac8 100644
--- a/.github/workflows/scale-test-node-throughput-gce.yaml
+++ b/.github/workflows/scale-test-node-throughput-gce.yaml
@@ -37,8 +37,8 @@ concurrency:
 env:
   # renovate: datasource=golang-version depName=go
   go_version: 1.22.2
-  # renovate: datasource=git-refs depName=https://github.com/kubernetes/perf-tests
-  cl2_version: a7c33443c28bbecb53584e3ca26422b3938e5ff2
+  # renovate: datasource=git-refs depName=https://github.com/kubernetes/perf-tests branch=master
+  cl2_ref: a7c33443c28bbecb53584e3ca26422b3938e5ff2
   # Adding k8s.local to the end makes kops happy-
   # has stricter DNS naming requirements.
   test_name: node-throughput
@@ -121,7 +121,7 @@ jobs:
         uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         with:
           repository: kubernetes/perf-tests
-          ref: ${{ env.cl2_version }}
+          ref: ${{ env.cl2_ref }}
           persist-credentials: false
           sparse-checkout: clusterloader2
           path: perf-tests

```

Note that I've changed `cl2_version` into `cl2_ref` to prevent the first regex from matching it as well (as it is looking for `.+_version`). Feel free to switch the suffix to something else if you prefer, everything should be fine as long as it is not matched by any other regex (it would likely still work, but also cause more noise due to the incorrect detection).

You can verify that it actually works by running (after changing one of the refs to an older version):

```
docker run -ti -e LOG_LEVEL=debug -e GITHUB_COM_TOKEN="$(gh auth token)" \
    -v /tmp:/tmp -v $(pwd):/usr/src/app \
    docker.io/renovate/renovate:full renovate --platform=local | tee renovate.log
```

As the logs will contain something along the lines of:

```
{
  "depName": "https://github.com/kubernetes/perf-tests",
  "currentValue": "master",
  "currentDigest": "82036a7e9bdb803b5b133c2d83cfec710819f9ab",
  "datasource": "git-refs",
  "replaceString": "# renovate: datasource=git-refs depName=https://github.com/kubernetes/perf-tests branch=master\n  cl2_ref: 82036a7e9bdb803b5b133c2d83cfec710819f9ab",
  "updates": [
    {
      "updateType": "digest",
      "newValue": "master",
      "newDigest": "a7c33443c28bbecb53584e3ca26422b3938e5ff2",
      "branchName": "renovate/all-github-action"
    }
  ],
  "packageName": "https://github.com/kubernetes/perf-tests",
  "versioning": "semver-coerced",
  "warnings": []
}
```

marseel replied:
I would consider dropping renovate - not sure if we want to have perf-tests updated constantly. It will make it harder for use to know if the performance difference comes from updating perf-tests or some change in cilium.

learnitall replied:
Ah yeah that's a good point @marseel. Thanks for the breakdown though @giorio94. I remember testing this locally and thinking it worked, but I don't think I looked for the JSON snippet you posted that had the `updates` key. TIL thank you!

giorio94 reviewed (changes requested):
Thanks! 

A comment regarding the renovate configuration which still doesn't work as expected (getting renovate to behave as you expect is pretty tricky, and requires some trial and error attempts), and a couple of additional nits inline.

learnitall committed:
ci: Update CL2 dep in scale test workflows

This commit updates the CL2 commit ref in the scale test workflows. The
goal with this update is to include changes from
kubernetes/perf-tests@5a36f2c78090144079ce9a3f26280ebf69eb5729.

This commit also adds a comment explaining why renovate is not used to
update the CL2 ref automatically, to help make it clear why this
functionality shouldn't be added in the future.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

giorio94 reviewed (changes requested):
Looks good to me, thanks!

(Requesting changes only because it still contains the temporary commit to be removed).

giorio94 reviewed (approved):
Thanks!

viktor-kurchenko reviewed (approved):
LGTM.
